odo/backends/url.py

```diff
@@ -0,0 +1,186 @@
+        with tmpfile(data.filename) as fn:
+            with open(fn, 'wb') as f:
+                f.write(raw.encode('utf-8'))
+            yield fn
+
+
+@discover.register((URL(CSV), URL(JSONLines)))
+def discover_url_line_delimited(c, lines=5, **kwargs):
+    """Discover CSV and JSONLines files from URL."""
+    with sample(c, lines=lines) as fn:
+        return discover(c.subtype(fn, **kwargs), **kwargs)
+
+
+types_by_extension = {'csv': CSV, 'json': JSONLines, 'txt': TextFile}
+
+
+@resource.register('ftp://.+', priority=16)
+@resource.register('http://.+', priority=16)
+@resource.register('https://.+', priority=16)
+def resource_url(uri, **kwargs):
+    path = os.path.basename(urlparse(uri).path)
+    try:
+        subtype = types_by_extension[ext(path)]
+    except KeyError:
+        subtype = type(resource(path))
+
+    return URL(subtype)(uri, **kwargs)
+
+
+@append.register(TextFile, URL(TextFile))
+@append.register(JSONLines, URL(JSONLines))
+@append.register(JSON, URL(JSON))
+@append.register(CSV, URL(CSV))
+def append_urlX_to_X(target, source, **kwargs):
+
+    with closing(urlopen(source.url)) as r:
+        chunk_size = 16 * source.chunk_size
+        with open(target.path, 'wb') as fp:
+            for chunk in iter(curry(r.read, chunk_size), ''):
+                fp.write(chunk)
+                return target
+
+
+@convert.register(Temp(TextFile), (Temp(URL(TextFile)), URL(TextFile)))
+@convert.register(Temp(JSONLines), (Temp(URL(JSONLines)), URL(JSONLines)))
+@convert.register(Temp(JSON), (Temp(URL(JSON)), URL(JSON)))
+@convert.register(Temp(CSV), (Temp(URL(CSV)), URL(CSV)))
+def url_file_to_temp_file(data, **kwargs):
+    fn = '.%s' % uuid.uuid1()
+    target = Temp(data.subtype)(fn, **kwargs)
+    return append(target, data, **kwargs)
+
+
+@convert.register(Temp(URL(TextFile)), (TextFile, Temp(TextFile)))
+@convert.register(Temp(URL(JSONLines)), (JSONLines, Temp(JSONLines)))
+@convert.register(Temp(URL(JSON)), (JSON, Temp(JSON)))
+@convert.register(Temp(URL(CSV)), (CSV, Temp(CSV)))
+def file_to_temp_url_file(data, **kwargs):
+    fn = '%s' % uuid.uuid1()
+    target = Temp(URL(getattr(data, 'persistent_type', type(data))))(fn, **kwargs)
+    return append(target, data, **kwargs)
+
+try:
+    from .hdfs import HDFS
+    from .aws import S3
+except ImportError:
+    pass
+else:
+    @append.register(S3(TextFile), URL(TextFile))
+    @append.register(S3(JSON), URL(JSON))
+    @append.register(S3(CSV), URL(CSV))
+    @append.register(S3(JSONLines), URL(JSONLines))
+    def other_remote_text_to_url_text(a, b, **kwargs):
+        raise MDNotImplementedError()
+
+
+    @append.register(HDFS(JSON), URL(JSON))
+    @append.register(HDFS(TextFile), URL(TextFile))
+    @append.register(HDFS(JSONLines), URL(JSONLines))
+    @append.register(HDFS(JSON), URL(JSON))
+    @append.register(HDFS(CSV), URL(CSV))
+    def append_url_to_hdfs(target, source, **kwargs):
+
+        path = os.path.basename(urlparse(source.url).path)
```

Comment:
can this be replaced by `source.filename`? it looks like you do this computation in the class initialization.